### ISSUE TITLE
Tips for some optimizations

### ISSUE THREAD

**Issue by fmassa** [jmhessel/fmpytorch#2]
Nice work with the code in this repo! I really liked to see the use of Cython with PyTorch!

One quick suggestion, for the second order comparison, it's possible to use some matrix-multiply operations in order to avoid the triple-nested for loop. Something like the following:
```python
# This code supposes PyTorch v0.2, as it uses broadcasting
class SecondOrderInteraction(nn.Module):
    def __init__(self, n_feats, n_factors):
        super(SecondOrderInteraction, self).__init__()
        self.v = nn.Parameter(torch.Tensor(n_feats, n_factors))
        self.v.data.uniform_(-0.01, 0.01)
        
    def forward(self, x):
        self.batch_size = x.size(0)
        all_interactions = torch.mm(self.v, self.v.t())
        pairwise = torch.bmm(x[:, :, None], x[:, None]) * all_interactions[None]
        mask = torch.ones(pairwise[0].size(), out=x.data.new()).triu(1)
        output = pairwise * Variable(mask)
        res = output.sum(1).sum(1,keepdim=True)
        return res
```
One advantage is that it works for CUDA out-of-the-box. But in other cases, using matrix-multiplies can become very memory-expensive, so the use of dedicated C/Cython code can be a much better approach.
Thanks!

**Comment by fmassa** [comment]
Closing because it's not an issue per-se, but just a comment.

**Comment by jmhessel** [comment]
Thanks for the observation and recoding! I'll give this a look. My original intention was to make an extremely naive implementation to make sure my gradients were correct. However, I'll definitely give this a closer look! :-)

**Comment by jmhessel** [comment]
Thanks again for the pure-torch implementation! As a quick follow up, I ran a few basic timing comparisons when comparing cython. Unfortunately, I didn't get a chance to run on the GPU, but here's some info:

For 200 input features, 25 hidden factors, batch size of 64, coupled with a linear/dropout layer, averaged over 1K batches:

Algo (CPU) | Time/batch
------------ | -------------
Super Naive (3 for loops) | ... a very long time...
Not naive (native torch from fmassa) | 38.1 ms
Cython | 3.48 ms


Obviously, different experimental settings will vary, and perhaps the GPU will speed things up. However, here's one datapoint! If folks are curious as to why I chose cython:

1. It gives easier control over exactly what memory is being handled (e.g., you don't need to instantiate the feature-by-feature interaction matrix)
2. I plan on adding sparse support at some point, and it's easier to deal with sparse tensors explicitly in cython, I think
3. I plan on adding arbitrary-order factorization machines at some point, and I'd expect there to be a fair number of tricky optimizations that need to be done to support that.
